Refactor(core): unify SSE computation methods

peppsac opened the pull request:
Until this PR each geometry type had its own SSE computation methods.
The goal of this PR is to provide a single method based on:
  - bounding-box (BB)
  - geometricError (in meters)
  - distance

The computation is simple:
  - transform camera in BB's basis
  - compute distance camera - BB
  - project geometricError on screen at distance

The last step produce a screen-space-error in pixels.

Geometric error values depend on the data type:
  - 3dtiles: the tileset defines a geometricError per node so we can use this value
  - pointcloud: PotreeConverter defines a spacing, which is the average distance
between points in each node. We use: geometricError = spacing / 2^depth
  - tiles: this PR adapted the existing formula to use consistent units

The second commit adds a SSE inspector debug tool:
![sse_inspector](https://user-images.githubusercontent.com/2198295/36847103-84733c3a-1d5d-11e8-961c-3136c82dc041.jpg)

red: computed SSE
yellow: texture resolution (only used for tile sse)
green: sse treshold

peppsac committed:
refactor(core): unify SSE computation methods

Until this commit each geometry type had its own SSE computation methods.
The goal of this commit is to provide a single method based on:
  - bounding-box (BB)
  - geometricError (in meters)
  - distance

The computation is simple:
  - transform camera in BB's basis
  - compute distance camera - BB
  - project geometricError on screen at distance

The last step produce a screen-space-error in pixels.

Geometric error values depend on the data type:
  - 3dtiles: the tileset defines a geometricError per node so we can use this value
  - pointcloud: PotreeConverter defines a spacing, which is the average distance
between points in each node. We use: geometricError = spacing / 2^depth
  - tiles: this commit adapted the existing formula to use consistent units

peppsac committed:
feat(core) add debug SSE inspector tool

peppsac commented:
Also note : this PR provides almost everything needed to compute the camera far / near plane correctly in realtime (near would be minimum distance from all updated elements, far would be the max)

gchoqueux commented on src/Core/Scheduler/Providers/PointCloudProvider.js:
```diff
@@ -83,7 +83,15 @@ function parseOctree(layer, hierarchyStepSize, root) {
+                        name: childname,
+                        baseurl: url,
+                        bbox: bounds,
+                        geometricError: layer.metadata.spacing / Math.pow(2, childname.length),
```
Evoked in the commit message, there is a reference that explains the calculation?

peppsac replied:
This is by construction: see https://github.com/peppsac/PotreeConverter/blob/master/PotreeConverter/src/PotreeWriter.cpp#L73

gchoqueux commented on src/Core/ScreenSpaceError.js:
```diff
@@ -0,0 +1,88 @@
+const m2 = new THREE.Matrix4();
+const m3 = new THREE.Matrix4();
+
+function compute(vector, matrix, camera, distance, _3d) {
```
`compute` what?

gchoqueux commented on src/Core/ScreenSpaceError.js:
```diff
@@ -0,0 +1,88 @@
+
+function compute(vector, matrix, camera, distance, _3d) {
+    const basis = [
+        new THREE.Vector3(vector.x, 0, 0),
```
`THREE.Vector3` to declare outside the function

gchoqueux commented on src/Core/ScreenSpaceError.js:
```diff
@@ -0,0 +1,88 @@
+
+function findBox3Distance(camera, box3, matrix) {
+    // TODO: can be cached
+    m.getInverse(matrix);
```
`matrix` it's OBB worldMatrix? you would cache in `OBB.worldMatrixInv`?

peppsac replied:
I'd prefer not to cache anything first (hence the TODO node).

I believe we need to rework the `notifyChange()` mechanism to allow more precise updates / cache invalidation.

gchoqueux commented on src/Core/ScreenSpaceError.js:
```diff
@@ -0,0 +1,88 @@
+
+    MODE_3D: 2,
+
+    computeFromBox3(camera, box3, matrix, geometricError, mode) {
```
there's no method to compute with sphere?

peppsac replied:
Not yet. As your other comment suggests I could move the sphere -> sse code from 3dtiles to here.

gchoqueux commented on src/Core/ScreenSpaceError.js:
```diff
@@ -0,0 +1,88 @@
+
+        const size = computeSizeFromGeometricError(box3, geometricError);
+
+        const sse = compute(size, matrix, camera, distance, mode == this.MODE_3D);
```
`compute` calculates a pixel projection size?

peppsac replied:
Yes, I'll change the function name.

gchoqueux commented on src/Core/ScreenSpaceError.js:
```diff
@@ -0,0 +1,88 @@
+}
+
+export default {
+    MODE_2D: 1,
```
could you explain the modes?

peppsac replied:
Yes I forgot to add a comment.
The idea here is: do we want to use the 2d bounding-box (x, y) or the 3d one.
For tiles we want to use the 2d one, while for other objects (pointcloud, 3dtiles) the 3d one makes more sense.

gchoqueux commented on src/Process/3dTilesProcessing.js:
```diff
@@ -225,25 +226,30 @@ const worldPosition = new THREE.Vector3();
         cameraLocalPosition.copy(camera.camera3D.position).sub(worldPosition);
-        node.distance = Math.max(0.0, node.boundingVolume.sphere.distanceToPoint(cameraLocalPosition));
+        const distance = Math.max(0.0, node.boundingVolume.sphere.distanceToPoint(cameraLocalPosition));
+        return camera.preSSE * (node.geometricError / distance);
```
why not place it in `ScreenSpaceError.js`

gchoqueux commented on src/Process/GlobeTileProcessing.js:
```diff
@@ -133,9 +96,14 @@ export function globeSubdivisionControl(minLevel, maxLevel, sseThreshold, maxDel
-        const sse = computeNodeSSE(context.camera, node);
-
-        return SSE_SUBDIVISION_THRESHOLD < sse;
+        node.sse = ScreenSpaceError.computeFromBox3(
```
why you replace the bounding sphere with OBBs?

peppsac replied:
Because the OBB is closer to the real volume of the `node` then a sphere.

gchoqueux commented on src/Process/GlobeTileProcessing.js:
```diff
@@ -164,6 +132,7 @@ export function globeSchemeTileWMTS(type) {
 }
 
 export function computeTileZoomFromDistanceCamera(distance, view) {
+    // TODO fixme
```
do you have an idea?

peppsac replied:
I didn't even try to fix this for now.

I'd rather make sure that the sse computation introduced in this PR is ok.

gchoqueux reviewed (changes requested):
review on first commit: `refactor(core): unify SSE computation methods `

peppsac committed:
Use tile size instead of half-size

peppsac committed:
review

peppsac commented:
PR updaded.
Note: the computation in `ScreenSpaceError.js` can probably be simplified.
But it's easier to review the non-optimized form and then as a last step simplify where possible.

peppsac commented:
Screenshots showing subdivision differences between master/PR :

Master:
![pr_off](https://user-images.githubusercontent.com/2198295/37044590-29c5b5a0-2164-11e8-9577-ddc091612688.jpg)
PR :
![pr_on](https://user-images.githubusercontent.com/2198295/37044592-29f7d1e8-2164-11e8-82af-5538c46a6594.jpg)

Or for pointclouds:
Master:
![pr_off2](https://user-images.githubusercontent.com/2198295/37045291-e13e9412-2165-11e8-9953-809f8822b81b.jpg)
PR:
![pr_on2](https://user-images.githubusercontent.com/2198295/37045306-ec6f91f6-2165-11e8-92a8-f516ab594fa8.jpg)

gchoqueux commented:
The tiles count between `camera's tilt` equals 0 and 85 is significant
But, qualitatively, it's too important for 0 and insufficient for 85.
can you balance the subdivision?

`tilt == 0`
![tilt == 0](https://user-images.githubusercontent.com/11291849/37050266-b8f6a248-2173-11e8-9b81-565aaaec2838.png)
`tilt == 85`
![tilt == 85](https://user-images.githubusercontent.com/11291849/37050322-d63b92dc-2173-11e8-9590-aae8a8a9f42c.png)

peppsac commented:
@gchoqueux : thanks for the feedback.

The current logic is: we project both the bounding box [x and y axis](https://github.com/iTowns/itowns/pull/674/files#diff-bc30a9362a318408471b5a8f16ce5b76R10) on the [screen](https://github.com/iTowns/itowns/pull/674/files#diff-bc30a9362a318408471b5a8f16ce5b76R31).
The SSE is then the [min of these sizes](https://github.com/iTowns/itowns/pull/674/files#diff-bc30a9362a318408471b5a8f16ce5b76R39).

Subdivision happens [if SSE](https://github.com/iTowns/itowns/pull/674/files#diff-989ef5eaaef422d31895b5f0e7eb2c7aR106) is > 256 (texture size) + layer.sseThreshold (5 by default).

The reason here for using the min is to prevent subdivision when the camera view vector is ~tangent to the surface but I agree that it's problematic in other case, such as your 2nd screenshot.

An alternative would be to keep SSE in x and y directions separately.
Then:
   - subdivide if min(sse.x, sse.y) > 256 or
   - if max(sse.x, sse.y) > 256 and min(sse.x, sse.y) > 128 (or another value that works)

peppsac committed:
Experiment globe/sse

peppsac commented:
@gchoqueux I pushed an example implementation of https://github.com/iTowns/itowns/pull/674#issuecomment-370891414

It breaks everything but the globe (because I only adapted GlobeTileProcessing). Can you give it a try and let me know what you think?

peppsac committed:
adapt sse

gchoqueux commented:
in `computeVectorSizeAtDistance` :
 * The box's diagonal may also be the largest projection?  This approach doesn't take it into account?
 * Could you use [Vector3.Project(camera)](https://github.com/mrdoob/three.js/blob/master/src/math/Vector3.js#L315)?
* the scale transformation isn't in the calculation?

peppsac commented:
> The box's diagonal may also be the largest projection? This approach doesn't take it into account?

This approach computes the sse for each axis.
For tiles, texture are mapped on the x/y axis, so I don't think it make sense to use the x/y diag.
For points, we could use the diagonal instead because the points are distributed in all directions in the bbox, but having the 3 sse values is enough it seems.

> Could you use Vector3.Project(camera)?

Nope. The `Vector3.Project` projects a world space vector, while I'm computing a [camera space vector](https://github.com/iTowns/itowns/pull/674/files#diff-bc30a9362a318408471b5a8f16ce5b76R32).
This vector is then moved at `distance` from the camera and then  projected on screen.

> the scale transformation isn't in the calculation?

It's used to compute the distance to camera.
For the size to projectd: we only the ratio of the box so the scale isn't relevant (because we scale the bbox to have its largest side being `geometricError`)

gchoqueux commented:
> so I don't think it make sense to use the x/y diag

I'm talking about the box's diagonal (in x, y and z)

> For tiles, texture are mapped on the x/y axis

DEM is mapped on z axis, not only x and y

>  while I'm computing a camera space vector.

yet you go through the world space when calculating (with `m3`)
the projection would be different with Vector3.Project?

peppsac commented:
> DEM is mapped on z axis, not only x and y

Can you compute a correct estimation of the DEM error in meters along the z axis for every tiles?
If you have a formula then yes we can use it as well.
Note that the current formula doesn't use the tile vertices count: it could make sense to use too but again, this needs a correct formula that works on all tiles.

> yet you go through the world space when calculating (with m3)

No I'm not. You misread the code: m3 is only the rotation part, not the full world matrix of the camera.

gchoqueux commented:
What is your formula to estimate error on the x and y axes?

You don't answer this question : the projection would be different with Vector3.Project?

peppsac commented:
> What is your formula to estimate error on the x and y axes?

Compare size in pixels of the tile on the screen to the texture size.

> You don't answer this question : the projection would be different with Vector3.Project?

Yes it would be diffrent. I detailled in my earlier comment https://github.com/iTowns/itowns/pull/674#issuecomment-371803340 the differences between Vector3.Project and my computations.

gchoqueux commented:
The texels are also on the box's diagonal (x, y, z), it can more revelant by example with box'size (100, 100, 1000)


> Yes it would be diffrent

I can't understand why there is a difference between the two, I'm going to look

peppsac committed:
update for z axis

peppsac committed:
fix rebase

peppsac committed:
Remove MODE_2D/3D option

peppsac commented:
Rebased. I also removed the `MODE_2D` / `MODE_3D` thing since it's not used anymore.

As far as I can tell this PR improves the SSE handling in most situations.

Regarding the diagonal: I still believe it isn't an issue. It's an corner case so I don't think it's worth complexifying the code to cover it. Btw, current master version SSE handling is quite bizarre anyway. E.g: in this screenshot the nearest tile in the middle of the screen is almost 600px wide.
 
![sse_bug_example](https://user-images.githubusercontent.com/2198295/38243368-9465ab52-3737-11e8-91b8-495c40b9e1ca.jpg)
You can try with:
 ```js
globeView.camera.camera3D.position.copy({x: 4490626.907379815, y: 506551.71291476686, z: 4500105.469154227})
globeView.camera.camera3D.rotation.copy({_x: -1.3092800928909716, _y: 0.4734022844137177, _z: -2.4859644421245957, _order: "XYZ"})
globeView.notifyChange(true);
```

mbredif commented on src/Core/ScreenSpaceError.js:
```diff
@@ -0,0 +1,96 @@
+    // Move camera position in box3 basis
+    // (we don't transform box3 to camera basis because box3 are AABB)
+    const pt = camera.camera3D.position
+        .clone(v).applyMatrix4(m);
```
this is wrong if m as some scaling (eigenvalues != 1)

mbredif commented on src/Core/ScreenSpaceError.js:
```diff
@@ -0,0 +1,96 @@
+        b.y = b.y * camera.height * 0.5;
+    }
+
+    return basis.map(b => b.length());
```
this looks like an approximation of some distance, which computation is optimized by considering axes independently. could you phrase mathematically the exact distance you are approximating ?

mbredif reviewed (commented):
some comments. could you define mathematically the sse?

peppsac commented:
> some comments. could you define mathematically the sse?

Thanks for the comments.

I'll come back to this PR soon, but for now I'm focused on the more recently opened PRs.
